Add EVP_MAC API

levitte opened the pull request:
We currently implement EVP MAC methods as EVP_PKEY methods.  This
change creates a separate EVP API for MACs, to replace the current
EVP_PKEY ones.

A note about this EVP API and how it interfaces with underlying MAC
implementations:

Other EVP APIs pass the EVP API context down to implementations, and
it can be observed that the implementations use the pointer to their
own private data almost exclusively.  The EVP_MAC API deviates from
that pattern by passing the pointer to the implementation's private
data directly, and thereby deny the implementations access to the
EVP_MAC context structure.  This change is made to provide a clearer
separation between the EVP library itself and the implementations of
its supported algorithm classes.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

levitte committed:
Add EVP_MAC API

We currently implement EVP MAC methods as EVP_PKEY methods.  This
change creates a separate EVP API for MACs, to replace the current
EVP_PKEY ones.

A note about this EVP API and how it interfaces with underlying MAC
implementations:

Other EVP APIs pass the EVP API context down to implementations, and
it can be observed that the implementations use the pointer to their
own private data almost exclusively.  The EVP_MAC API deviates from
that pattern by passing the pointer to the implementation's private
data directly, and thereby deny the implementations access to the
EVP_MAC context structure.  This change is made to provide a clearer
separation between the EVP library itself and the implementations of
its supported algorithm classes.

levitte commented:
This is a start.  There's more work to do, such as moving the current EVP_PKEY implementations to use the EVP_MAC API instead, tests, refinement, and documentation.

t-j-h commented:
Initial thoughts.

Why does init not take a key? Using a ctrl to provide it seems unusual. "oneshot" as a name does not fit current API naming elsewhere does it?

Why void * for the key? Are you expecting something other than raw bytes? If so it should be typed appropriately. If not it should be unsigned char *.

t-j-h commented:
And that approach pretty much ignores the current generic API under EVP_PKEY_new_mac_key which is what you are looking at replacing. That is where the "pattern" should come from IMHO.

And what I'm curious about is what happens to be wrong with the current interface in this area. It is used and is workable.  It lets you use EVP_DigestSign/EVP_DigestVerify as the interface for both signatures and MACs in a unified manner. 

This approach was added back in 2007 - i.e. it is a 1.0.0 API - when SteveH modeled MACs as effectively being signatures.

I think looking at the original initial low level HMAC interface isn't the right place to draw from in terms of the model.

I'm also curious as to the driver behind wanting to replace the existing EVP_PKEY model approach.

levitte commented:
I'll have a closer look at `EVP_PKEY_new_mac_key`.  I did notice it, but thought it didn't tell the whole story.

As for the driver, it's mostly clarity.  A MAC is kinda sorta a digest and yet not.  This is not the first time that we find ourselves being quite confused by the current construct.

Remember that with 1.0.x, we had *everything* in public headers, which lead to many attempts to retrofit new features into already existing structures.

I'm sure @mattcaswell has a few choice words on this as well.  What got me immediately started now was the repeated desire for a MAC API, mentioned in #7387

InfoHunter commented:
> There's more work to do, such as moving the current EVP_PKEY implementations to use the EVP_MAC API instead, tests, refinement, and documentation.

Right. And if this PR is done, then people can take what they are interested in to move to the new APIs concurrently...

mattcaswell commented:
> And what I'm curious about is what happens to be wrong with the current interface in this area. It is used and is workable. It lets you use EVP_DigestSign/EVP_DigestVerify as the interface for both signatures and MACs in a unified manner.

It is a gross perversion of what an EVP_PKEY is supposed to be and is totally counter intuitive and non-obvious.

The state of this API was mentioned recently by @slontis as a justification for not using EVP when it was easier and simpler to just use HMAC directly:
 
https://github.com/openssl/openssl/pull/6779#discussion_r216897292

MACs are not signatures. Nor are they digests. Yes they have some similarities to both but they are a different primitive and deserve their own dedicated high level API.

levitte commented:
> MACs are not signatures. Nor are they digests. Yes they have some similarities to both but they are a different primitive and deserve their own dedicated high level API.

In an operations based API (which our current API is *not*), it would be expressed in terms of "MACs can be used for..." rather than "MACs are..."

levitte commented:
I have now beefed this up quite a bit.  The EVP_MAC API is *done*, the quick adaptation of existing code is *done*.  All current tests pass, and because the current EVP_PKEY_METHOD implementations have been recoded to use EVP_MAC, it means that the new code is automatically tested as well.

Still lacking: documentation, and seeing if I can make a more generic MAC interface to enhance / replace `EVP_PKEY_new_CMAC_key`...  I frankly don't see a reason why we should have something so specialised at the EVP level.

levitte commented:
Description got updated too, with the text from the first commit.

InfoHunter commented:
I am a little confused about this. For MAC algorithms like poly1305, should that be used by calling the newly added `EVP_MAC_*` interfaces?

levitte commented:
Depends, really.  If you want to use it as a signing mechanism, then it's still going to be through EVP_PKEY, at least as far as I can deduce.  Please note that this is still work in progress.

slontis commented:
Could you add some tests that use the new API's directly? Maybe the key should be an init parameter (that can be NULL).

InfoHunter commented:
> Could you add some tests that use the new API's directly? 

I plan to make a new PR (actually it's already underway...) based on this PR, which should include some test cases I think, but I found it's a little tricky to figure out the dependencies between PRs...

InfoHunter commented:
> I plan to make a new PR (actually it's already underway...)

To make it clearer, that PR will contain an implementation for poly1305 based on the new MAC APIs.

levitte commented:
I will add tests, yeah.

As for the key passed in init, that's more difficult. I would like the EVP_MAC implementations to be fairly minimal, and the underlying low level implementation may require that the associated crypto algorithm be set before the key (HMAC_Init works that way, for example), so if we send the key via init, then we must also send the associated crypto algorithm, which is an EVP_MD for HMAC and an EVP_CIPHER for CMAC. At that point, we're getting ourselves into a type safety issue, and since there's no telling what the future will give, it seemed saner to set all the associated data via ctrl.

Another option would be for every EVP_MAC to save the key away and use it later, like hm_pmeth.c does. I would rather not have to force every EVP_MAC to implement that kind of mechanism. 

I hope that rationale is good enough.

InfoHunter commented:
> Could you add some tests that use the new API's directly? Maybe the key should be an init parameter (that can be NULL).

Please take a reference to the test case in #7459

levitte commented:
Yup, will do...  for the moment, I'm gonna battle the memleak to put all memleaks to shame (I've never seen that many lines displayed before)

InfoHunter commented:
> (I've never seen that many lines displayed before)

Neither do I ;-)

levitte commented:
Memory leak gone...  however, the API isn't quite as complete as I wish for.  Fair warning, you'll have to work a bit more on your Poly1305 work to fit with what's happening with this.

tmshort commented on crypto/cmac/cm_meth.c:
```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2010-2016 The OpenSSL Project Authors. All Rights Reserved.
```
```suggestion
 * Copyright 2010-2018 The OpenSSL Project Authors. All Rights Reserved.
```

tmshort commented:
So, what happens with the EVP_PKEY interface for HMAC, CMAC (and eventually Poly1305)? This is explicitly stated as a replacement, but the EVP_PKEY to these algorithms remain.
If both interfaces remain, are we concerned about having two (three in some cases) ways to do the same thing? I'm not sure it makes these algorithms easier to use, but actually harder to use, given the multiple ways something can be done.

levitte commented:
I'm looking quite a lot at that, and while I stated **replacement** at the beginning, I'm not going to take away the EVP_PKEY interface at this stage, especially since there may be some usage out there.  However, what I'm currently doing is to create a EVP_PKEY -> EVP_MAC bridge, which means that the diverse MAC specific EVP_PKEY_METHOD implementations are going away (in separate PRs).

levitte commented:
Please do note WIP

levitte committed:
Add automatic initializations support for EVP_MAC objects

levitte committed:
Add generic EVP_PKEY_METHOD for EVP_MACs

The MAC EVP_PKEY implementations are currently implemented for each
MAC.  However, with the EVP_MAC API, only one such implementation is
needed.

This implementation takes into account the differences between HMAC
and CMAC implementations, and observes that all other current MAC
implementations seem to follow the HMAC model.

levitte committed:
Adapt test/evp_test.c to deal with available EVP_MACs

If a MAC isn't available as an EVP_MAC, the MAC test falls back to the
corresponding EVP_PKEY method.

levitte committed:
Add convenience functions EVP_str2ctrl() and EVP_hex2ctrl()

These functions are generalizations of EVP_PKEY_CTX_str2ctrl() and
EVP_PKEY_CTX_hex2ctrl().  They will parse the value, and then pass the
parsed result and length to a callback that knows exactly how to pass
them on to a main _ctrl function, along with a context structure
pointer.

levitte commented:
Note that this PR only adds the API itself and the initializing infrastructure.

Other PRs will do the actual port of the different MAC algos to EVP_MAC and to the EVP_PKEY->EVP_MAC bridge.

levitte commented:
HMAC port in #7483

levitte commented:
CMAC port in #7484

levitte commented:
And @InfoHunter is working on a Poly1305 port in #7459 (it will need a bit of work...  sorry for pulling the rug from under you like that, @InfoHunter)

paulidale commented:
@tmshort my thought would be that the EVP_MAC interface becomes canonical, the other interfaces being rewritten to use the EVP_MAC.  At some later point the other interfaces are deprecated and later removed or otherwise changed to legacy status.

InfoHunter commented:
It's very reasonable to separate the HMAC and CMAC out of this PR

levitte commented:
I've added documentation (thanks for the reminder, @kroeckx), and fixed up `crypto/evp/pkey_mac.c` after having a run through with SipHash (PR coming up!).

I think this is as ready as I can get it at this stage.

levitte commented:
I don't understand what's happening here...

```
ccache clang  -I. -Icrypto/include -Iinclude -fPIC -pthread -m64 -Wa,--noexecstack -Qunused-arguments -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wmissing-prototypes -Wstrict-prototypes -Wshadow -Wformat -Wtype-limits -Wundef -Werror  -Wswitch-default -Wno-parentheses-equality -Wno-language-extension-token -Wno-extended-offsetof -Wconditional-uninitialized -Wincompatible-pointer-types-discards-qualifiers -Wmissing-variable-declarations -Wno-unknown-warning-option -Wall -O3 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-1.1\"" -DNDEBUG -DOPENSSL_API_COMPAT=0x10100000L -D__NO_STRING_INLINES -MMD -MF crypto/evp/libcrypto-shlib-pkey_mac.d.tmp -MT crypto/evp/libcrypto-shlib-pkey_mac.o -c -o crypto/evp/libcrypto-shlib-pkey_mac.o crypto/evp/pkey_mac.c
crypto/evp/pkey_mac.c:338:23: error: no previous extern declaration for non-static variable 'FAKE_pkey_meth' [-Werror,-Wmissing-variable-declarations]
const EVP_PKEY_METHOD FAKE_pkey_meth = {
                      ^
```

Isn't this declaration enough?

``` C
const EVP_PKEY_METHOD FAKE_pkey_meth;
const EVP_PKEY_METHOD FAKE_pkey_meth = {
```

https://github.com/openssl/openssl/pull/7393/commits/68639242f83017a4797b6852de89f012f441e5bc#diff-ffb1b6c2ed0dd0134c6d5cfe3b97d47aR337

levitte commented:
D'oooooh....  staring myself blind, I didn't notice the missing `extern`

levitte commented:
... and it builds!

Time for reviewers to step forward :-)

kroeckx commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,235 @@
+EVP_get_macbyname(), EVP_get_macbynid() and EVP_get_macbyobj() return
+the request MAC implementation, if it exists in the object database,
+otherwise B<NULL>.
+
```
Maybe an example is useful

levitte replied:
Ah, thanks, I was thinking I should, but forgot.

kroeckx commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,235 @@
+
+EVP_MAC_nid() return the numeric identity of the given MAC implementation.
+
+EVP_MAC_name() return the numeric identity of the given MAC implementation.
```
I'm going to guess it returns a name, not a number.

levitte replied:
Yes.  Will fix.

levitte replied:
Fixed

kroeckx commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,235 @@
+=head2 Registration functions
+
+EVP_add_mac() adds the MAC implementation C<mac> to the internal
+object database.
```
I'm not sure what the use case of this is. Do we somehow let applications defined their own EVP_MAC?

levitte replied:
Good question, I actually haven't thought that far.  This is, however, a function that gets used, see the MAC reimplementations I've provided in PRs #7483, #7484 and #7494.

Do you want a full MAC creation API *now*, or could that be an expansion for another PR?

kroeckx replied:
Making it an internal function instead?

levitte replied:
Done, will come in a push soon

kroeckx commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,235 @@
+EVP_MAC_CTX_new() creates a new context for the MAC type C<mac>.
+EVP_MAC_CTX_new_id() creates a new context for the numerical MAC
+identity <nid>.  The created context can then be used with most other
+functions described here.
```
I wonder why we use an NID here, and not something like EVP_HMAC(), which at least seems to be the old style. In any case, I think we should at least document which NIDs/OBJs/Names are supported.

levitte replied:
You're right regarding documentation, I will add that in the relevant PRs

levitte replied:
I have documented the NIDs in PRs #7483, #7484 and #7494.  Will document the names as well.

InfoHunter replied:
I am also curious about why not using the old style EVP_xxx() function to create a new context...

levitte replied:
Essentially, `EVP_HMAC()` would be a shortcut for `EVP_get_macbynid(EVP_MAC_HMAC)` (which ends up calling `EVP_get_macbyname` with the name from the ASN.1 object found with the nid).  It avoids the cycles to find the desired MAC, but is also a pattern that is giving us trouble that I think is best avoided, for now.

If it's *really* desirable, I can add them...  or someone else can.  Is it?

kroeckx commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,235 @@
+diverse controls.  Implementations may vary on this, some of them may
+do their initialization in the controls themselves.  However, this
+should be called at the very least before calling EVP_MAC_update() and
+EVP_MAC_final().
```
This reads as, you should call it, but we're not telling you when, only that it needs to be before update, it might need to be before or after ctrl(), and you need to figure that out on your own. I wonder why it needs to be called at all, why isn't EVP_MAC_CTX_new() doing it already?

I don't think we should support reusing, so I don't see a need for this init function.

levitte replied:
EVP_MAC_CTX_new only creates and initializes the MAC level context, but not the underlying computation algo if there is one. EVP_MAC_init does the latter, after that algo has been specified.

So the call sequence would typically be:

EVP_MAC_CTX_new
EVP_MAC_ctrl (enough times)
EVP_MAC_init
EVP_MAC_update (one or more)
EVP_MAC_final

kroeckx replied:
Note that EVP_PKEY_derive_init() does it just the other way around, init first. And as #7497 points out, the init function might not do what you expect to reuse it.

levitte replied:
EVP_PKEY is a mixture of things, so it's quite hard to draw conclusions from one quick example.

In the MAC case, the driving question is how to initialize an underlying computation algorithm without knowing what that algorithm will be.

levitte replied:
Thinking a bit more about it, there are definitely controls that can be called after init.  It all depends on what they affect.

levitte replied:
I take that back...  all controls defined so far affect things that must be in place before the MAC is initialized.

kroeckx replied:
What do you think about internally calling this init function in update or final if it wasn't called yet?

kroeckx commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,235 @@
+EVP_MAC_final().
+
+EVP_MAC_reset() resets the computation for the given context.  This
+may not be supported by the underlying implementation.
```
I think this should get removed.

kroeckx replied:
I think it either it should always be supported, or should be removed.

kroeckx replied:
You can also wonder what the difference between init and reset is.

levitte replied:
Supporting it always could be done, actually.
Is it worth it or is it better to simply removed?

I need help to decide

levitte replied:
Init resets the key, i.e whatever initial calculations the actual MAC implementation does from the given key will be recalculated. The reset (resume in CMAC speak) should start over with the key material already calculated from last init. It's meant to save cycles.

As far as I know, only HMAC and CMAC supports this. For any other algo, reset will probably just do an init from a saved key.

richsalz replied:
If we don't know that the API is needed, don't implement it seems like a reasonable starting rule.

levitte replied:
Okie, I think I'll remove this.  For HMAC, it simply zeroes the HMAC context, so it's the same as zallocing the HMAC context again, except for the allocation part...

CMAC is undocumented and we use `CMAC_resume` *even less* than `HMAC_CTX_reset`, so it's as good a reason as any to simply turn a blind eye...

kroeckx commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,235 @@
+
+EVP_MAC_CTX_copy() makes a deep copy of the C<src> context to the
+C<dest> context.  The C<dest> context I<must> have been created before
+calling this function.
```
Maybe a dup() function is more useful?

levitte replied:
```
: ; grep dup include/openssl/evp.h 
EVP_MD *EVP_MD_meth_dup(const EVP_MD *md);
EVP_CIPHER *EVP_CIPHER_meth_dup(const EVP_CIPHER *cipher);
EVP_PKEY_CTX *EVP_PKEY_CTX_dup(EVP_PKEY_CTX *ctx);
: levitte@lapdog:~/gitwrk/openssl.net/official/master
 * this file except in compliance with the License.  You can obtain a copy
int EVP_MD_meth_set_copy(EVP_MD *md, int (*copy)(EVP_MD_CTX *to,
int (*EVP_MD_meth_get_copy(const EVP_MD *md))(EVP_MD_CTX *to,
/* cipher has its own additional copying logic */
int EVP_CIPHER_CTX_copy(EVP_CIPHER_CTX *out, const EVP_CIPHER_CTX *in);
__owur int EVP_MD_CTX_copy_ex(EVP_MD_CTX *out, const EVP_MD_CTX *in);
__owur int EVP_MD_CTX_copy(EVP_MD_CTX *out, const EVP_MD_CTX *in);
int EVP_ENCODE_CTX_copy(EVP_ENCODE_CTX *dctx, EVP_ENCODE_CTX *sctx);
int EVP_PKEY_copy_parameters(EVP_PKEY *to, const EVP_PKEY *from);
void EVP_PKEY_asn1_copy(EVP_PKEY_ASN1_METHOD *dst,
                             int (*param_copy) (EVP_PKEY *to,
void EVP_PKEY_meth_copy(EVP_PKEY_METHOD *dst, const EVP_PKEY_METHOD *src);
void EVP_PKEY_meth_set_copy(EVP_PKEY_METHOD *pmeth,
                            int (*copy) (EVP_PKEY_CTX *dst,
void EVP_PKEY_meth_get_copy(const EVP_PKEY_METHOD *pmeth,
                            int (**pcopy) (EVP_PKEY_CTX *dst,
```

(an interesting thing to note is that EVP_CIPHER has a copy function for its context, but calls the method function pointer `dup` :wink:)

I think I'll stick with the copy pattern...

kroeckx replied:
I really dislike the copy pattern. I'm sure most of those copy functions have various bugs in them and are very hard to get correct, and probably none of them are correct. This is part of the reusing pattern that I want to get rid of.

kroeckx replied:
The options for me are either providing a dup function, or removing this function.

levitte replied:
I think I recall seeing calls to copy when running through tests. I'll do a run with a profiler to see if that's still the case or if it went away somehow.

levitte replied:
As it is right now, we cannot get away from a copy function, because there are variants of key passing down to an MAC (a CMAC in my testing case) from `EVP_DigestSignInit` that requires copying a `EVP_MAC_CTX`.  `EVP_PKEY_new_CMAC_key` is the function used to create that `EVP_MAC_CTX`.  So getting rid of the copy function means digging very deep in this MAC-over-PKEY stuff.  I'm not ready to do that, and *certainly* not as part of this PR.

It would of course be desirable to get rid of MAC-over-PKEY, but since it's part of the current API and something people have learned to use, we can't...  which means that for this PR, the copy function stays.

kroeckx commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,235 @@
+
+=item B<EVP_MAC_CTRL_SET_ENGINE>
+
+=item B<EVP_MAC_CTRL_SET_MD>
```
The above 2 are duplicates documented below

levitte replied:
No.  These lines are items, the others are text within the block starting with these items.  You should look at the document with `perldoc` to see how it renders.

richsalz replied:
You might want to use comma-separated items in a single "item" entry as was done in many other manpages.

kroeckx commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,235 @@
+
+=over 4
+
+=item B<EVP_MAC_CTRL_SET_KEY>
```
Should we have functions for those controls instead?

We normally do this using a define that call the control function, but I would prefer real functions.

levitte replied:
I absolutely hate the proliferation of little functions for this, especially since they've appeared a bit hodge podge over time, with a certain variety in naming, which at has at least me quite confused over what's actually going on.  Having to trace my way through such calls to the point of diziness is no fun.

So no, I have no desire whatsoever to add such functions.  If someone else really wants them, it will be up to them to create the PR.

kroeckx commented:
blake2 can also be used as a MAC using a key, but as far as I remember, we didn't implement the key/mac part. This might be a good opportunity to add that.

levitte commented:
> blake2 can also be used as a MAC using a key, but as far as I remember, we didn't implement the key/mac part. This might be a good opportunity to add that.

Sure.  That can be done as a separate PR, just like there are for HMAC, CMAC, SipHash and Poly1305

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -0,0 +1,197 @@
+
+    if (src->data != NULL && (dst->data = dst->meth->copy(src->data)) == NULL) {
+        EVPerr(EVP_F_EVP_MAC_CTX_COPY, ERR_R_MALLOC_FAILURE);
+        return 0;
```
I finally came to realize that this isn't a good implementation, so I'm changing the `copy` EVP_MAC function to behave the same way as other copying functions do.

levitte committed:
fixup! Add EVP_MAC API

levitte committed:
fixup! Add EVP_MAC API

levitte committed:
fixup! Add EVP_MAC API

InfoHunter commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,338 @@
+different types and with different underlying algorithms if there are
+any.
+
+MACs are a bit complex insofar that some of them use other other
```
It seems the word 'other' is duplicated...

levitte replied:
Thanks. Will fix.

InfoHunter commented on crypto/evp/mac_lib.c:
```diff
@@ -0,0 +1,194 @@
+    EVP_MAC_CTX *ctx = OPENSSL_zalloc(sizeof(EVP_MAC_CTX));
+
+    /* Dummy check that there is a freeing function as well */
+    mac->free(NULL);
```
why this is needed?

levitte replied:
Strictly speaking, it isn't. That's some left over debugging code I forgot... I will remove.

paulidale commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,338 @@
+the memory pointed at by C<out>, and sets its size in the B<size_t>
+the C<poutlen> points at.  If C<out> is B<NULL>, then no computation
+is made.  To figure out what the output length will be and allocate
+space for it dynamically, simply call with C<out> being B<NULL> and
```
I'd prefer a separate interface to return the length rather than the pass NULL hack used in many places.  For API compatibility, we're need to support the latter but still.....

levitte replied:
You mean, like `EVP_MAC_size`?

levitte replied:
Either way, you make a good point.  There are already other patterns being broken here, so why not this as well...  question is how many patterns we should break in one go.

Toss a coin and decide.  I'm ambivalent and can go either way.

paulidale replied:
I am hopeful that the KDF API will avoid the pass NULL and get the size back syndrome.  So yes, `EVP_MAC_size` sounds good to me.  I'll cope either way of course, but one API for one purpose seems more natural than overloading things.

paulidale commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,338 @@
+C<poutlen> pointing at a valid location, then allocate space and make
+a second call with C<out> pointing at the allocated space.
+
+EVP_MAC_ctrl() is used to manipulate or get information on aspects of
```
There has to be a better way.  I cannot see it yet.

levitte replied:
The pattern we've had is to add a gazillion little functions that are used to tell some object "hey, take this!", and that end up calling control functions.  Basically, it's hard to avoid for functionality that has diverse backends, and MACs are quite diverse!  So we end up with a model like `ioctl`s...

paulidale commented on doc/man3/EVP_MAC.pod:
```diff
@@ -0,0 +1,338 @@
+These will set the MAC flags to the given numbers.  Some MACs do not
+support this option.
+
+=item B<EVP_MAC_CTRL_SET_ENGINE>
```
Do we require this given providers are the future...

levitte replied:
When (and if, there's still an "if") we get there, this will need to be redone anyway...  This PR deals with "now".

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -0,0 +1,194 @@
+    /* Dummy check that there is a freeing function as well */
+    mac->free(NULL);
+
+    if (ctx == NULL
```
The source has changed, and I didn't look for collapsible lines.  Fixed.

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -0,0 +1,194 @@
+    return ctx;
+}
+
+static void evp_mac_ctx_cleanup(EVP_MAC_CTX *ctx)
```
The now removed `EVP_MAC_reset` used it.  Cleaned up now.

levitte committed:
fixup! Add EVP_MAC API

paulidale commented on crypto/evp/mac_lib.c:
```diff
@@ -0,0 +1,194 @@
+    return ctx->meth->final(ctx->data, out);
+}
+
+int EVP_MAC_ctrl(EVP_MAC_CTX *ctx, int cmd, ...)
```
I've no objection, it seemed odd that's all.

paulidale commented on crypto/evp/names.c:
```diff
@@ -178,3 +206,47 @@ void EVP_MD_do_all_sorted(void (*fn) (const EVP_MD *md,
+static void do_all_mac_fn(const OBJ_NAME *nm, void *arg)
+{
+    struct doall_mac *dc = arg;
+    if (nm->alias)
```
Blank line between declaration and body?

levitte replied:
Fixed

paulidale commented on crypto/evp/pkey_mac.c:
```diff
@@ -0,0 +1,369 @@
+static int pkey_mac_init(EVP_PKEY_CTX *ctx)
+{
+    MAC_PKEY_CTX *hctx;
+    /* EVP_PKEY_id(EVP_PKEY_CTX_get0_pkey(ctx)) */
```
Blank line again.

levitte replied:
Fixed (actually, removed the comment)

paulidale reviewed (approved):
Two nits.

levitte commented:
Fixed the two nits, then merged.

Thanks for the reviews, everyone!  Much appreciated.

f842b6b2a4d7cbb9d22e4605c502b73f25bb6a7b Add convenience functions EVP_str2ctrl() and EVP_hex2ctrl()
2bdb4af50364121a5f0e47024e9f71e1a6025fcf Adapt test/evp_test.c to deal with available EVP_MACs
5e55159b3adbb30482992e8fa8621e47d331d012 Add generic EVP_PKEY_METHOD for EVP_MACs
0145dd324e8fcfd2c0dfe296c12586101f0cf3b9 Add automatic initializations support for EVP_MAC objects
567db2c17d4ea8a0164d7abd8aed65b7a634bb40 Add EVP_MAC API
